### PR TITLE
[IA-2967] Adding autopauseEnabled value to runtime

### PR DESCRIPTION
**Pull request by cstrandw**
RELATED TO: https://github.com/DataBiosphere/terra-ui/pull/2815

- Changed logic in calculateAutopauseThreshold to accommodate no autopauseEnabled value being passed in on update

Testing:
- Ran tests locally
- Tested manually through the UI and checked corresponding database values for validity
- See https://github.com/DataBiosphere/terra-ui/pull/2815 for other testing there

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/RuntimeRoutesTestJsonCodec.scala [hunk 1]
```diff
@@ -81,7 +81,7 @@ object RuntimeRoutesTestJsonCodec {
     "startUserScriptUri",
     "runtimeConfig",
     "userJupyterExtensionConfig",
-    "autopause",
+    "autopauseEnabled",
```
Comment (Qi77Qi) [inline core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/RuntimeRoutesTestJsonCodec.scala]: this will be breaking change for clients. Terra UI isn't the only client for Leonardo. So I'd prefer we keep the previous name

File: http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20220207_autopause_enabled.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+<databaseChangeLog logicalFilePath="leonardo" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+    <changeSet logicalFilePath="leonardo" author="cstrandw" id="autopause_enabled">
+        <addColumn tableName="CLUSTER">
+            <column name="autopauseEnabled" type="boolean" defaultValue="false">
+            </column>
+        </addColumn>
+    </changeSet>
+
+    <changeSet id="set autopauseEnabled to false when autopauseThreshold is 0, else set to true" author="cstrandw">
+        <sql>UPDATE CLUSTER SET autopauseEnabled = TRUE WHERE autopauseThreshold != 0</sql>
```
Comment (Qi77Qi) [inline http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20220207_autopause_enabled.xml]: I don't think we need this field. Like I said in slack, `autopauseThreshold` essentially represents all cases already
* None --> autopause disable
* Some(v) --> autopause enabled, and its threshold is “v”

Comment (cstrandw) [reply http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20220207_autopause_enabled.xml]: I was talking with @gpcarr and @jdcanas about whether we should keep the current implementation or add the enabled value to the database, and we agreed that it would best practice to add it instead of relying on deriving the value every time

Comment (Qi77Qi) [reply http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20220207_autopause_enabled.xml]: > deriving the value every time

where will be be "deriving" the value?

The downside of having both `autopauseEnabled` and `autopauseThreshold` is you'll like have invalid states. For instance, you have `autopauseEnabled` being `None`, and `autopauseThreshold` being `Some(v)`....this will be invalid. The two fields aren't independent of each other

**Comment by cstrandw** [comment]
> Seeing this formatting change, I just wanted to check if you've been running `sbt scalafmtAll` on your Terminal :)

Yes, that's what I did

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala [hunk 3]
```diff
@@ -103,6 +103,7 @@ final case class GetRuntimeResponse(id: Long,
                                     startUserScriptUri: Option[UserScriptPath],
                                     errors: List[RuntimeError],
                                     userJupyterExtensionConfig: Option[UserJupyterExtensionConfig],
+                                    autopause: Boolean,
```
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala]: what is this boolean for?

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala [hunk 4]
```diff
@@ -35,6 +35,7 @@ final case class Runtime(id: Long,
                          errors: List[RuntimeError],
                          dataprocInstances: Set[DataprocInstance],
                          userJupyterExtensionConfig: Option[UserJupyterExtensionConfig],
+                         autopause: Boolean,
```
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala]: don't think this is needed?

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala (hunk 3 above)
Comment (cstrandw) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala]: I removed the autopause column from the database, but kept it in the models so that it would keep the logic I previously implemented.

**Comment by Qi77Qi** [comment]
used `git reset --merge HEAD~1` to remove the bad merge

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 5]
```diff
@@ -29,6 +29,7 @@ import java.sql.SQLDataException
 import java.time.Instant
 import scala.concurrent.ExecutionContext
 
+// TODO decide if we need to add autopause?
```
Comment (kyuksel) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Is this comment needed?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala [hunk 6]
```diff
@@ -1014,12 +1022,13 @@ object RuntimeServiceInterp {
...
     autopause match {
       case None =>
-        autoFreezeConfig.autoFreezeAfter.toMinutes.toInt
+        autoPauseOffValue
       case Some(false) =>
         autoPauseOffValue
       case _ =>
-        if (autopauseThreshold.isEmpty) autoFreezeConfig.autoFreezeAfter.toMinutes.toInt
-        else Math.max(autoPauseOffValue, autopauseThreshold.get)
+        if (autopauseThreshold.isEmpty) 30 //TODO what behavior do we want when enabled but no threshold
```
Comment (kyuksel) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala]: Magic number alert :) If we default to a value (like `30` here), it should probably come from a config, or at the least it should be defined somewhere higher.

Comment (cstrandw) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala]: Planning to change! Talked to Qi about it earlier

File: core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/RuntimeRoutesTestJsonCodec.scala [hunk 7]
```diff
@@ -95,7 +95,7 @@ object RuntimeRoutesTestJsonCodec {
       x.startUserScriptUri,
       x.runtimeConfig,
       x.userJupyterExtensionConfig,
-      x.autopause,
+      x.autopause.getOrElse(false),
```
Comment (Qi77Qi) [inline core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/RuntimeRoutesTestJsonCodec.scala]: you don't need `.getOrElse(false)` here

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala [hunk 8]
```diff
@@ -82,7 +82,7 @@ object UpdateRuntimeConfigRequest {
 
 final case class UpdateRuntimeRequest(updatedRuntimeConfig: Option[UpdateRuntimeConfigRequest],
                                       allowStop: Boolean,
-                                      updateAutopauseEnabled: Option[Boolean],
+                                      updateAutopause: Option[Boolean],
```
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala]: is this rename intended?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 9]
```diff
@@ -758,6 +758,7 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
       startUserScriptUri = clusterRecord.startUserScriptUri,
       errors = errors map clusterErrorQuery.unmarshallClusterErrorRecord,
       userJupyterExtensionConfig = extensionQuery.unmarshallExtensions(userJupyterExtensionConfig),
+      autopause = clusterRecord.autopauseThreshold != autoPauseOffValue,
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: think we can remove `autopause` from `Runtime` model?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/RuntimeRoutes.scala [hunk 10]
```diff
@@ -547,6 +547,7 @@ object RuntimeRoutes {
       ("jupyterStartUserScriptUri", x.startUserScriptUri.asJson),
       ("errors", x.errors.asJson),
       ("userJupyterExtensionConfig", x.userJupyterExtensionConfig.asJson),
+      ("autopause", x.autopause.asJson),
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/RuntimeRoutes.scala]: so I think we can probably remove `autopause` from `GetRuntimeResponse` as well, and just adding this field in the encoder to make UI easier.

you can have
```
val autoPauseEnabled = x.autopauseThreshold != autoPauseOffValue

Json.obj(
...
("autopause", autoPauseEnabled.asJson)
...
```

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala (hunk 3 above)
Comment (cstrandw) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala:106]: @Qi77Qi I want the autopause value to be returned in the response for the UI, so shouldn't I keep this value here and not remove it? Or do you want me to refactor the UI side to work only with the threshold value?

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala (hunk 8 above)
Comment (cstrandw) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala]: The rename was intended to help this match the corresponding variable, which is simply autopause.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/RuntimeRoutes.scala (hunk 10 above)
Comment (cstrandw) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/RuntimeRoutes.scala:550]: Where is the encoder? Still getting familiar with the codebase

**Review by Qi77Qi — APPROVED** [review]
the syntax can be updated a bit...but otherwise looks good

```
autopause match {
      case Some(false) =>
        autoPauseOffValue
      case _ =>
        autopauseThreshold match {
          case Some(v) => v
          case None    => autoFreezeConfig.autoFreezeAfter.toMinutes.toInt
        }
    }
```
